base/rational.jl

```diff
@@ -129,12 +129,299 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
+                        !mantissa_is_one &
+                        (romo_is_ntaf | (romo_is_ntte & !mantissa_is_even))
+                    )
+                )
+            )
+        )
+            # Round up
+
+            # Assuming `T` is wide enough and there's no overflow.
+            mantissa += true
+
+            # We need to decrease the bit width in case it increased.
+            mantissa_carry = ispow2(mantissa)
+            mantissa >>= mantissa_carry
+        elseif romo_is_to_nearest & to_nearest_is_tied & mantissa_is_one
+            # Mantissa is one, which means the precision is also
+            # one. Be consistent with the `BigFloat` behavior, for
+            # example: `BigFloat(3) == BigFloat(3.0) == 4`.
+            mantissa_carry = true
+        end
+    end
+
+    # `mantissa` should now again be exactly `requested_bit_width` bits
+    # wide.
+
+    exp = bit_shift_2 - bit_shift_1 + mantissa_carry
+
+    (
+        mantissa = mantissa,
+        exponent = exp,
+        is_exact = result_is_exact,
+    )
+end
+
+# `num`, `den` are positive integers. `requested_bit_width` is the
+# requested floating-point precision and must be positive. `T` is the
+# integer type that we'll be working with mostly, it needs to be wide
+# enough.
+function rational_to_float_components(
+    num::Integer,
+    den::Integer,
+    requested_bit_width::Integer,
+    ::Type{T},
+    romo::RoundingMode,
+) where {T<:Integer}
+    (false < requested_bit_width) || error("nonpositive bit width")
+
+    # Factor out powers of two
+    trailing_zeros_num = trailing_zeros(num)
+    trailing_zeros_den = trailing_zeros(den)
+    num >>= trailing_zeros_num
+    den >>= trailing_zeros_den
+
+    c = rational_to_float_components_impl(
+        num, den, requested_bit_width, T, romo,
+    )
+
+    (
+        mantissa = c.mantissa,
+        exponent = c.exponent + trailing_zeros_num - trailing_zeros_den,
+        is_exact = c.is_exact,
+    )
+end
+
+function rational_to_float_impl(
+    to_float::C,
+    ::Type{<:Integer},
+    x::Rational{Bool},
+    ::Integer,
+) where {C<:Union{Type,Function}}
+    n = numerator(x)
+    if iszero(denominator(x))
+        if iszero(n)
+            to_float(NaN)  # 0/0
+        else
+            to_float(Inf)  # 1/0
+        end
+    else
+        # n/1 = n
+        to_float(n)
+    end
 end
 
+to_int8_if_bool(n::Bool) = Int8(n)
+to_int8_if_bool(n::Integer) = n
+
+# Assuming the wanted rounding mode is round to nearest with ties to
+# even.
+#
+# `requested_precision` must be positive.
+function rational_to_float_impl(
+    to_float::C,
+    ::Type{T},
+    x::Rational,
+    requested_precision::Integer,
+) where {C<:Union{Type,Function},T<:Integer}
+    s = Int8(sign(numerator(x)))
+    a = abs(x)
+
+    num = to_int8_if_bool(numerator(a))
+    den = to_int8_if_bool(denominator(a))
+
+    # Handle special cases
+    num_is_zero = iszero(num)
+    den_is_zero = iszero(den)
```

Comment:
I know, the issue was that `Bool` `trailing_zeros`, as far as I remember.

Reply:
Won't that already be handled by the specialization for `Rational{Bool}` in the function just above this one?

Reply:
This was done with types like `Rational{Union{Bool,Int}}` in mind. The idea was that, although neither the numerator or the denominator will be of type `Bool` with the current constructors (as discussed in the recently closed issue):
1. this seems like something that might change in the future, as the constructor implementation changes
2. the compiler doesn't (I guess) know about the constructor implementation, so the conditional conversion into `Int8` might help its inference.

Reply:
Ok, but that's still an unnecessary conversion and at least a comment explaining the reasoning would have been good. 

Julia is a dynamic language, so special casing for some hypothetical (when the more generic `iszero` works just fine even for the `Rational{Union{Bool, Int}}` case) is not really good style. Without a concrete justification why `Int8` specifically is better here, this is just additional noise when reading the code. Please keep the implementation simple.

Reply:
It's also not that the compiler doesn't know about the constructor implementation (which, while doing the conversion now, _is not guaranteed to_) it's that the type of the elements _cannot ever be_ inferred as anything other than `Union{Bool, Int8}` semantically.

Reply:
Actually, I was wrong when I said "neither the numerator or the denominator will be of type `Bool` with the current constructors":

```julia-repl
julia> typeof(numerator(Rational{Union{Bool,Int8}}(true, true)))
Bool
```

So we do need to handle this case.

> when the more generic `iszero` works just fine

No, as I wrote above, the conversion from `Bool` is because `Bool` doesn't have `trailing_zeros`:

```julia-repl
julia> trailing_zeros(false)
ERROR: MethodError: no method matching trailing_zeros(::Bool)
```

The `trailing_zeros` function is used in `rational_to_float_components`. I could make a wrapper around `trailing_zeros` that also supports `Bool`, then we wouldn't need to use `to_int8_if_bool`. Do you think that would be an improvement?

Reply:
I think worrying about mixed-type `Rational` conversion in the same implementation focused on correctness of the existing conversion of homogenous types is ill-advised.

Reply:
I'm not sure what do you propose?

Reply:
Do you want types like `Rational{Union{Bool,Int8}}` to just be unsupported?

Reply:
No, I'm proposing to have a slow, but correct fallback implementation for these kinds of odd types, that doesn't rely on things like `trailing_zeros` (what would `trailing_zeros(false)` be?). The optimized/performance sensitive version should be concerned with a concrete type only, like the one handling `Rational{Bool}` above.

Reply:
> I'm proposing to have a slow, but correct fallback implementation for these kinds of odd types

So a new code path? But we still need the same algorithms...

> that doesn't rely on bitlevel inspection

This PR relies on bit-level inspection anyway, though. Bit-level inspection is actually quite essential.

> what would `trailing_zeros(false)` be?

We know `n` is positive at the places where this would be used, so for `n isa Bool` we necessarily have `n === true`, so there are zero trailing zeros. So we could potentially use something like this:

```julia
# `n` is positive
trailing_zeros_supporting_true(n::Integer) = trailing_zeros(n)
trailing_zeros_supporting_true(::Bool) = trailing_zeros(UInt8(true))
```

However, I think this could be a bit confusing, so IMO staying with something like `to_int8_if_bool` might be the cleanest approach.

Reply:
> This PR relies on bit-level inspection anyway, though. Bit-level inspection is actually quite essential.

Bit-level inspection doesn't make sense for `Bool` though. There is only one bit. `sizeof(Bool)` only reports 1 because the other 7 bits are padding. Hence, any `Rational` potentially holding `Bool` should be dealt with differently (not to mention that your `bit_width` gives `bit_width(false) === 0`). 

If conversion is the only option to not bloat the codebase, eagerly converting `Rational{Union{Bool, <:Integer}}` to a concrete `Rational` (which should always be possible losslessly) instead of inline ad-hoc conversion should be preferred, for example via explicit `promote` and reconstruction of the `Rational`, before handing it off to the function handling homogenous types:

```
function (::Type{F})(x::Rational{T}) where {F<:AbstractFloat, T <: Integer}
      nr = Rational(promote(denominator(x), numerator(x))...)
      # pass `nr` into `rational_to_float` for the actual conversion
end
```

Reply:
For example, the `gcd` function supports mixed argument types just fine, by `promote`ing them to a common type, and only then forwarding to the specialized implementation:

https://github.com/JuliaLang/julia/blob/0a05a5b05d382bdf38da7fdd0c069dff8bdc072c/base/intfuncs.jl#L143

Please refer to the manual section on [Conversion & Promotion](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/) for more information on how this system works.

Reply:
> your bit_width gives `bit_width(false) === 0`

That's OK and as intended.

> If conversion is the only option to not bloat the codebase, eagerly converting `Rational{Union{Bool, <:Integer}}` to a concrete `Rational` [...]

OK, on it.